File: src/easy/node-depth/c-sharp/stack.cs
using System;
using System.Collections.Generic;

/*
Average case: O(n) time | O(h) space (when the tree is balanced)
Worst case  : O(n) time | O(n) space ("single-branch" tree stack space)
Where N is the number of nodes, and H is the height of the Binary tree
*/
public class Program {
    public static int NodeDepths (BinaryTree root) {
        int sum = 0;
        Stack<TreeLevel> store = new Stack<TreeLevel> ();

        store.Push (new TreeLevel (root, 0));

        while (store.Count > 0) {
            TreeLevel current = store.Pop ();
            BinaryTree node = current.root;
            int depth = current.depth;

            if (node != null) {
                sum += depth;
                store.Push (new TreeLevel (node.left, depth + 1));
                store.Push (new TreeLevel (node.right, depth + 1));
            }
        }

        return sum;
    }

    public class TreeLevel {
        public BinaryTree root;
        public int depth;

        public TreeLevel (BinaryTree root, int depth) {
            this.root = root;
            this.depth = depth;
        }
    }

    public class BinaryTree {
        public int value;
        public BinaryTree left;
        public BinaryTree right;

        public BinaryTree (int value) {
            this.value = value;
            left = null;
            right = null;
        }
    }
}

File: src/binary-trees/extreme/sum-all-dephts/go/stack.go
package main

type BinaryTree struct {
	Value       int
	Left, Right *BinaryTree
}

// For the average case where the tree is balanced
// O(n.log(n)) time | O(h) space
// where N is the number of nodes and H the height of the Binary Tree
func AllKindsOfNodeDepths(root *BinaryTree) int {
	sumOfDepths := 0
	stack := []*BinaryTree{root}
	var node *BinaryTree
	for len(stack) > 0 {
		node, stack = stack[len(stack)-1], stack[:len(stack)-1]
		if node == nil {
			continue
		}
		sumOfDepths += nodeDepths(node, 0)
		stack = append(stack, node.Left)
		stack = append(stack, node.Right)
	}
	return sumOfDepths
}

func nodeDepths(node *BinaryTree, depth int) int {
	if node == nil {
		return 0
	}
	return depth + nodeDepths(node.Left, depth+1) + nodeDepths(node.Right, depth+1)
}
